File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/advatec/AdvatecWebhookListener.kt
package com.fccmiddleware.edge.adapter.advatec

import com.fccmiddleware.edge.logging.AppLogger
import io.ktor.http.ContentType
import io.ktor.http.HttpStatusCode
import io.ktor.server.cio.CIO
import io.ktor.server.engine.EmbeddedServer
import io.ktor.server.engine.embeddedServer
import io.ktor.server.request.header
import io.ktor.server.request.receiveText
import io.ktor.server.response.respondText
import io.ktor.server.routing.post
import io.ktor.server.routing.routing
import kotlinx.coroutines.CancellationException
import java.time.Instant
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Advatec Receipt webhook listener (ADV-3.2).
 *
 * Listens for Receipt webhook callbacks from the Advatec EFD device on a
 * configurable port. Validates the webhook token (X-Webhook-Token header or
 * ?token= query parameter), then enqueues valid payloads for the adapter to
 * drain and normalize.
 *
 * Always returns HTTP 200 OK — Advatec retry behaviour is unknown (AQ-7),
 * so we avoid triggering potential retries on validation errors.
 *
 * Thread safety: The incoming queue is a lock-free [ConcurrentLinkedQueue].
 * The server runs on CIO (coroutine I/O) threads — no Android main thread blocking.
 */
class AdvatecWebhookListener(
    /** Port for the embedded HTTP server to listen on. */
    private val listenPort: Int,
    /** Site code to stamp on received payloads. */
    private val siteCode: String,
    /** Expected webhook token for authentication. Null/blank = skip validation. */
    private val webhookToken: String?,
) {
    companion object {
        private const val TAG = "AdvatecWebhookListener"

        /** Maximum queued receipts before dropping (back-pressure safety). */
        const val MAX_QUEUE_SIZE = 10_000
    }

    /** Thread-safe queue of raw JSON payloads received from the Advatec EFD. */
    private val incomingQueue = ConcurrentLinkedQueue<ReceivedPayload>()

    /** Atomic counter tracking queue size to avoid TOCTOU race. */
    private val queueCount = AtomicInteger(0)

    /** Whether the listener is currently running. */
    private val running = AtomicBoolean(false)

    /** Embedded Ktor server instance. Null when stopped. */
    private var server: EmbeddedServer<*, *>? = null

    /**
     * Whether the listener is currently accepting connections.
     */
    val isRunning: Boolean
        get() = running.get()

    /**
     * Number of payloads currently queued and not yet drained by the adapter.
     */
    val queueSize: Int
        get() = incomingQueue.size

    /**
     * Starts the embedded HTTP server.
     *
     * Idempotent — calling start on an already-running listener is a no-op.
     * The server binds to 0.0.0.0:[listenPort] and accepts POSTs on /api/webhook/advatec.
     *
     * @return true if the server started (or was already running), false on bind failure
     */
    fun start(): Boolean {
        if (running.getAndSet(true)) {
            AppLogger.d(TAG, "Already running on port $listenPort")
            return true
        }

        return try {
            val srv = embeddedServer(CIO, port = listenPort) {
                routing {
                    post("/api/webhook/advatec") {
                        handleWebhookRequest(call)
                    }
                    // Also accept on catch-all for flexibility (Advatec webhook URL may vary)
                    post("/api/webhook/advatec/") {
                        handleWebhookRequest(call)
                    }
                }
            }
            srv.start(wait = false)
            server = srv
            AppLogger.i(TAG, "Started on port $listenPort (path: /api/webhook/advatec)")
            true
        } catch (e: Exception) {
            AppLogger.e(TAG, "Failed to start on port $listenPort: ${e.message}")
            running.set(false)
            false
        }
    }

    /**
     * Stops the embedded HTTP server and clears the incoming queue.
     *
     * Idempotent — calling stop on an already-stopped listener is a no-op.
     * Waits up to 2 seconds for graceful shutdown.
     */
    fun stop() {
        if (!running.getAndSet(false)) {
            return
        }

        try {
            server?.stop(gracePeriodMillis = 1000, timeoutMillis = 2000)
            AppLogger.i(TAG, "Stopped")
        } catch (e: Exception) {
            AppLogger.w(TAG, "Error during shutdown: ${e.message}")
        } finally {
            server = null
            incomingQueue.clear()
            queueCount.set(0)
        }
    }

    /**
     * Drains all queued payloads (up to [maxCount]) and returns them.
     *
     * Non-blocking. Returns an empty list if no payloads are available.
     * The adapter calls this during its fetch cycle to collect pushed receipts.
     *
     * @param maxCount Maximum number of payloads to drain in one call
     * @return List of received payloads, oldest first
     */
    fun drainQueue(maxCount: Int = 200): List<ReceivedPayload> {
        val result = mutableListOf<ReceivedPayload>()
        repeat(maxCount) {
            val item = incomingQueue.poll() ?: return result
            queueCount.decrementAndGet()
            result.add(item)
        }
        return result
    }

    // -----------------------------------------------------------------------
    // Private — Request handling
    // -----------------------------------------------------------------------

    /**
     * Handles an incoming HTTP POST from the Advatec EFD.
     *
     * Validation steps:
     * 1. Read raw JSON body
     * 2. Validate webhook token (X-Webhook-Token header or ?token= param)
     * 3. Enqueue the raw JSON for later normalization
     * 4. Return HTTP 200 OK
     *
     * Always returns 200 — Advatec retry behaviour is unknown (AQ-7).
     */
    private suspend fun handleWebhookRequest(call: io.ktor.server.application.ApplicationCall) {
        try {
            // Step 1: Read raw JSON body
            val rawJson = call.receiveText()
            if (rawJson.isBlank()) {
                AppLogger.w(TAG, "Received empty body")
                call.respondText("OK", ContentType.Text.Plain, HttpStatusCode.OK)
                return
            }

            // Step 2: Validate webhook token
            if (!webhookToken.isNullOrBlank()) {
                val providedToken = call.request.header("X-Webhook-Token")
                    ?: call.request.queryParameters["token"]

                if (providedToken.isNullOrBlank() || providedToken != webhookToken) {
                    AppLogger.w(TAG, "Rejected request with invalid or missing webhook token")
                    // Still return 200 to avoid leaking token validation info
                    call.respondText("OK", ContentType.Text.Plain, HttpStatusCode.OK)
                    return
                }
            }

            // Step 3: Enqueue for adapter processing (with back-pressure guard)
            val newCount = queueCount.incrementAndGet()
            if (newCount > MAX_QUEUE_SIZE) {
                queueCount.decrementAndGet()
                AppLogger.w(TAG, "Queue full ($MAX_QUEUE_SIZE) — dropping receipt")
                call.respondText("OK", ContentType.Text.Plain, HttpStatusCode.OK)
                return
            }

            val payload = ReceivedPayload(
                rawJson = rawJson,
                siteCode = siteCode,
                receivedAt = Instant.now(),
            )
            incomingQueue.add(payload)
            AppLogger.d(TAG, "Enqueued receipt webhook (queue size=$newCount, body=${rawJson.length} bytes)")

            // Step 4: Return 200 OK
            call.respondText("OK", ContentType.Text.Plain, HttpStatusCode.OK)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            AppLogger.e(TAG, "Error handling webhook request: ${e::class.simpleName}: ${e.message}")
            try {
                call.respondText("OK", ContentType.Text.Plain, HttpStatusCode.OK)
            } catch (_: Exception) {
                // Response already sent or connection closed — ignore
            }
        }
    }
}

/**
 * A raw receipt payload received via the Advatec webhook.
 *
 * Stored in the [AdvatecWebhookListener]'s queue until drained by the adapter
 * for normalization.
 */
data class ReceivedPayload(
    /** Complete raw JSON body from the webhook POST request. */
    val rawJson: String,
    /** Site code stamped on receipt. */
    val siteCode: String,
    /** When the webhook was received by the listener. */
    val receivedAt: Instant,
)

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Advatec/AdvatecWebhookListener.cs
using System.Net;
using System.Text;
using FccDesktopAgent.Core.Adapter.Common;
using Microsoft.Extensions.Logging;

namespace FccDesktopAgent.Core.Adapter.Advatec;

/// <summary>
/// Minimal HTTP listener for receiving Advatec Receipt webhook callbacks (ADV-3.1).
/// <para>
/// Listens on a configurable port for POST /api/webhook/advatec.
/// Validates the X-Webhook-Token header against the configured token.
/// Always returns 200 OK to Advatec (retry behaviour unknown — AQ-7) to avoid data loss.
/// Delivers received payloads to a callback delegate for processing.
/// </para>
/// </summary>
public sealed class AdvatecWebhookListener : IDisposable
{
    /// <summary>
    /// Delegate invoked when a valid webhook payload is received.
    /// The implementation should not throw; exceptions are caught and logged.
    /// </summary>
    /// <param name="payload">The raw webhook envelope ready for normalization.</param>
    /// <param name="ct">Cancellation token.</param>
    public delegate Task WebhookReceivedHandler(RawPayloadEnvelope payload, CancellationToken ct);

    private const string WebhookPath = "/api/webhook/advatec";
    private const string TokenHeader = "X-Webhook-Token";

    private readonly int _port;
    private readonly string _siteCode;
    private readonly string? _webhookToken;
    private readonly ILogger<AdvatecWebhookListener> _logger;

    private HttpListener? _listener;
    private CancellationTokenSource? _cts;
    private Task? _listenTask;
    private WebhookReceivedHandler? _handler;

    /// <summary>
    /// Creates a new Advatec webhook listener.
    /// </summary>
    /// <param name="port">TCP port to listen on (e.g. 8091).</param>
    /// <param name="siteCode">Site code to stamp on received payloads.</param>
    /// <param name="webhookToken">
    /// Expected value of the X-Webhook-Token header. If null or empty, token validation is skipped.
    /// </param>
    /// <param name="logger">Logger instance.</param>
    public AdvatecWebhookListener(
        int port,
        string siteCode,
        string? webhookToken,
        ILogger<AdvatecWebhookListener> logger)
    {
        _port = port;
        _siteCode = siteCode;
        _webhookToken = webhookToken;
        _logger = logger;
    }

    /// <summary>
    /// Whether the listener is currently running and accepting connections.
    /// </summary>
    public bool IsListening => _listener?.IsListening ?? false;

    /// <summary>
    /// Starts the webhook listener on the configured port.
    /// </summary>
    /// <param name="handler">
    /// Callback invoked for each valid webhook payload. Must be set before calling Start.
    /// </param>
    /// <param name="ct">Cancellation token to stop the listener.</param>
    /// <exception cref="InvalidOperationException">Thrown if the listener is already running.</exception>
    public void Start(WebhookReceivedHandler handler, CancellationToken ct = default)
    {
        if (_listener is not null)
            throw new InvalidOperationException("Advatec webhook listener is already running");

        _handler = handler ?? throw new ArgumentNullException(nameof(handler));
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);

        _listener = new HttpListener();
        _listener.Prefixes.Add($"http://+:{_port}{WebhookPath}/");

        try
        {
            _listener.Start();
        }
        catch (HttpListenerException ex)
        {
            _logger.LogError(ex,
                "Advatec webhook listener failed to start on port {Port} (path: {Path}). "
                + "On Windows, you may need to grant a URL ACL for this port.",
                _port, WebhookPath);
            _listener = null;
            _cts.Dispose();
            _cts = null;
            throw;
        }

        _logger.LogInformation(
            "Advatec webhook listener started on port {Port} (path: {Path})",
            _port, WebhookPath);

        _listenTask = AcceptLoopAsync(_cts.Token);
    }

    /// <summary>
    /// Gracefully stops the webhook listener.
    /// </summary>
    public async Task StopAsync()
    {
        if (_cts is not null)
        {
            await _cts.CancelAsync();
        }

        if (_listener is not null)
        {
            try
            {
                _listener.Stop();
                _listener.Close();
            }
            catch (ObjectDisposedException)
            {
                // Already disposed.
            }

            _listener = null;
        }

        if (_listenTask is not null)
        {
            try
            {
                await _listenTask;
            }
            catch (OperationCanceledException)
            {
                // Expected on shutdown.
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Advatec webhook listener accept loop ended with error");
            }

            _listenTask = null;
        }

        _logger.LogInformation("Advatec webhook listener stopped");
    }

    public void Dispose()
    {
        _cts?.Cancel();
        try
        {
            _listener?.Stop();
            _listener?.Close();
        }
        catch (ObjectDisposedException)
        {
            // Already disposed.
        }

        _cts?.Dispose();
        _listener = null;
        _cts = null;
    }

    // ── Private methods ──────────────────────────────────────────────────────

    private async Task AcceptLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            HttpListenerContext context;
            try
            {
                context = await _listener!.GetContextAsync().WaitAsync(ct);
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch (HttpListenerException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch (ObjectDisposedException)
            {
                break;
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Advatec webhook listener: error accepting connection");
                continue;
            }

            // Process each request in the background so the accept loop stays responsive.
            _ = ProcessRequestAsync(context, ct);
        }
    }

    private async Task ProcessRequestAsync(HttpListenerContext context, CancellationToken ct)
    {
        var response = context.Response;

        try
        {
            var request = context.Request;

            // Only accept POST requests.
            if (!string.Equals(request.HttpMethod, "POST", StringComparison.OrdinalIgnoreCase))
            {
                _logger.LogDebug("Advatec webhook: rejected {Method} request (only POST accepted)", request.HttpMethod);
                await WriteResponse(response, 405, "Method Not Allowed");
                return;
            }

            // Validate X-Webhook-Token header.
            if (!string.IsNullOrEmpty(_webhookToken))
            {
                var providedToken = request.Headers[TokenHeader];

                // Also check query parameter ?token= as a fallback (Advatec may configure webhook URL with token param).
                if (string.IsNullOrEmpty(providedToken))
                {
                    providedToken = request.QueryString["token"];
                }

                if (string.IsNullOrEmpty(providedToken) ||
                    !string.Equals(providedToken, _webhookToken, StringComparison.Ordinal))
                {
                    _logger.LogWarning("Advatec webhook: rejected request with invalid or missing webhook token");
                    // Still return 200 to avoid leaking information about token validation.
                    await WriteResponse(response, 200, "OK");
                    return;
                }
            }

            // Read the request body.
            string body;
            using (var reader = new System.IO.StreamReader(request.InputStream, Encoding.UTF8))
            {
                body = await reader.ReadToEndAsync(ct);
            }

            if (string.IsNullOrWhiteSpace(body))
            {
                _logger.LogWarning("Advatec webhook: received empty body");
                await WriteResponse(response, 200, "OK");
                return;
            }

            _logger.LogDebug("Advatec webhook: received payload ({Length} bytes)", body.Length);

            // Always return 200 OK immediately (Advatec retry behaviour unknown — AQ-7).
            await WriteResponse(response, 200, "OK");

            // Build the raw payload envelope and deliver to the handler.
            var envelope = new RawPayloadEnvelope(
                FccVendor: "Advatec",
                SiteCode: _siteCode,
                RawJson: body,
                ReceivedAt: DateTimeOffset.UtcNow);

            if (_handler is not null)
            {
                try
                {
                    await _handler(envelope, ct);
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "Advatec webhook: handler failed for payload (body length={Length})", body.Length);
                    // Do not re-throw -- the 200 is already sent.
                }
            }
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            _logger.LogError(ex, "Advatec webhook: unhandled error processing request");

            // Best-effort 200 response to avoid potential retries.
            try
            {
                await WriteResponse(response, 200, "OK");
            }
            catch
            {
                // Response may already be sent or connection may be broken.
            }
        }
        finally
        {
            try
            {
                response.Close();
            }
            catch
            {
                // Ignore close errors.
            }
        }
    }

    private static async Task WriteResponse(HttpListenerResponse response, int statusCode, string body)
    {
        response.StatusCode = statusCode;
        response.ContentType = "text/plain; charset=utf-8";

        var buffer = Encoding.UTF8.GetBytes(body);
        response.ContentLength64 = buffer.Length;

        await response.OutputStream.WriteAsync(buffer);
        await response.OutputStream.FlushAsync();
    }
}
